src/pages/WorkspaceDetails/index.tsx

```diff
@@ -77,92 +73,88 @@ export class WorkspaceDetails extends React.PureComponent<Props, State> {
-        return WorkspaceDetailsTab.Devfile;
+      if (searchParam.has('tab') && searchParam.get('tab') === WorkspaceDetailsTab.DEVFILE) {
+        return WorkspaceDetailsTab.DEVFILE;
       }
     }
+    return WorkspaceDetailsTab.OVERVIEW;
+  }
 
-    return WorkspaceDetailsTab.Overview;
+  public componentDidMount(): void {
+    this.unregisterLocationCallback = this.props.history.listen(location => {
+      const activeTabKey = this.getActiveTabKey(location.search);
+      if (activeTabKey !== this.state.activeTabKey) {
+        this.setState({ activeTabKey });
+      }
+    });
   }
 
-  public componentDidUpdate(): void {
-    const activeTabKey = this.getActiveTabKey();
-    if (this.state.activeTabKey !== activeTabKey) {
-      this.setState({ activeTabKey });
+  public componentWillUnmount() {
+    if (this.unregisterLocationCallback) {
+      this.unregisterLocationCallback();
     }
-    if (this.props.workspace && (WorkspaceStatus[this.props.workspace?.status] === WorkspaceStatus.STOPPED)) {
+  }
+
+  public componentDidUpdate(): void {
+    if (this.props.workspace && this.props.workspace?.isStopped) {
       this.setState({ hasWarningMessage: false });
     }
   }
 
-  private handleDiscardChanges(): void {
-    if (this.state.activeTabKey === WorkspaceDetailsTab.Devfile) {
+  private handleDiscardChanges(pathname: string): void {
+    if (this.state.activeTabKey === WorkspaceDetailsTab.DEVFILE) {
       this.editorTabPageRef.current?.cancelChanges();
-    } else if (this.state.activeTabKey === WorkspaceDetailsTab.Overview) {
+    } else if (this.state.activeTabKey === WorkspaceDetailsTab.OVERVIEW) {
       this.overviewTabPageRef.current?.cancelChanges();
     }
 
-    const tabIndex = this.state.clickedTabIndex;
-    this.setState({ hasDiscardChangesMessage: false, activeTabKey: tabIndex });
+    if (pathname.startsWith('/workspace/')) {
+      const tabIndex = this.state.clickedTabIndex;
+      const searchParams = new window.URLSearchParams(this.props.history.location.search);
+      searchParams.set('tab', tabIndex as WorkspaceDetailsTab);
+      this.props.history.location.search = searchParams.toString();
+      this.props.history.push(this.props.history.location);
+    } else {
+      this.props.history.push(pathname);
+    }
   }
 
-  private handleCancelChanges(): void {
-    this.setState({ hasDiscardChangesMessage: false });
+  private hasUnsavedChanges(): boolean {
+    if (this.state.activeTabKey === WorkspaceDetailsTab.DEVFILE) {
+      return this.editorTabPageRef.current?.state.hasChanges || !this.editorTabPageRef.current?.state.isDevfileValid;
+    } else if (this.state.activeTabKey === WorkspaceDetailsTab.OVERVIEW) {
+      return this.overviewTabPageRef.current?.hasChanges === true;
+    }
+    return false;
   }
 
   public render(): React.ReactElement {
-    const { workspace, workspacesLink } = this.props;
+    const { workspace, workspacesLink, history } = this.props;
```

Comment:
It works just fine for the described issue but seems I found one more case when it still does not work, it does not work in case of incorrect YAML, just add any symbol after `:`, or remove space between any key: value pair:
![Screenshot_20210524_152939](https://user-images.githubusercontent.com/5887312/119348095-169dd400-bca5-11eb-8111-79334335018f.png)

Then you're able to switch to overview without ask to discard changes.
When you return back to devfile editor - your invalid yaml is there, if you haven't applied changes on Overview page.

Reply:
I've created a separate issue to handle that since it's not bug introduced of this PR https://github.com/eclipse/che/issues/19873